h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/matrix/SparseMatrixFactory.java

```diff
@@ -0,0 +1,366 @@
+package hex.tree.xgboost.matrix;
+
+import hex.DataInfo;
+import ml.dmlc.xgboost4j.java.DMatrix;
+import ml.dmlc.xgboost4j.java.XGBoostError;
+import water.H2O;
+import water.LocalMR;
+import water.MrFun;
+import water.fvec.Chunk;
+import water.fvec.Frame;
+import water.fvec.Vec;
+import water.util.ArrayUtils;
+
+import static hex.tree.xgboost.matrix.MatrixFactoryUtils.setResponseAndWeight;
+import static water.MemoryManager.*;
+import static water.MemoryManager.malloc4;
+
+public class SparseMatrixFactory {
+
+    public static DMatrix csr(
+        Frame f, int[] chunksIds, Vec weightsVec, Vec responseVec, // for setupLocal
+        DataInfo di, float[] resp, float[] weights)
+        throws XGBoostError {
+
+        SparseMatrixDimensions sparseMatrixDimensions = calculateCSRMatrixDimensions(f, chunksIds, weightsVec, di);
+        SparseMatrix sparseMatrix = allocateCSRMatrix(sparseMatrixDimensions);
+
+        Vec.Reader[] vecs = new Vec.Reader[f.numCols()];
+        for (int i = 0; i < vecs.length; ++i) {
+            vecs[i] = f.vec(i).new Reader();
+        }
+        Vec.Reader weightsReader = (weightsVec != null) ? weightsVec.new Reader() : null;
+        Vec.Reader responseReader = responseVec.new Reader();
+
+        int actualRows = initializeFromChunkIds(
+            f, chunksIds, vecs, weightsReader,
+            di, sparseMatrix, sparseMatrixDimensions,
+            responseReader, resp, weights);
+
+        return toDMatrix(sparseMatrix, sparseMatrixDimensions, actualRows, di);
+    }
+
+    public static DMatrix csr(Chunk[] chunks, int weight, int respIdx, // for MR task
+        DataInfo di, float[] resp, float[] weights) throws XGBoostError {
+
+        SparseMatrixDimensions sparseMatrixDimensions = calculateCSRMatrixDimensions(chunks, di, weight);
+        SparseMatrix sparseMatrix = allocateCSRMatrix(sparseMatrixDimensions);
+
+        int actualRows = initializeFromChunks(
+            chunks, weight,
+            di, sparseMatrix._rowHeaders, sparseMatrix._sparseData, sparseMatrix._colIndices,
+            respIdx, resp, weights);
+        return toDMatrix(sparseMatrix, sparseMatrixDimensions, actualRows, di);
+    }
+
+    private static DMatrix toDMatrix(SparseMatrix sm, SparseMatrixDimensions smd, int actualRows, DataInfo di) throws XGBoostError {
+        DMatrix trainMat = new DMatrix(sm._rowHeaders, sm._colIndices, sm._sparseData,
+            DMatrix.SparseType.CSR, di.fullN(), actualRows + 1,
+            smd._nonZeroElementsCount);
+        assert trainMat.rowNum() == actualRows;
+        return trainMat;
+    }
+    
+    private static class NestedArrayPointer {
```

Comment:
Encapsulating this in a separate class is great  for readability.

The reason this was not previously nested was simple: I had no time to check if JVM will inline the increment operation, as it could lead to a potential overhead (this part was aimed to be as high performance as possible). Can we somehow check this ? Or do we consider the overhead negligible in general ? I've tried it locally with `-XX:+UnlockDiagnosticVMOptions -XX:+PrintInlining` and it does not show this method gets inlined.

Here is the log:
[log.txt](https://github.com/h2oai/h2o-3/files/3236957/log.txt)

Reply:
I have no thought about that, but unless there is a good reason I think readability wins.